File: functions/deleteSensorV2/s3RemoverV2.js
import { S3Client, GetObjectCommand, PutObjectCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';

const s3Client = new S3Client(
  process.env.IS_LOCAL
    ? {
        region: 'us-east-2',
        endpoint: 'http://host.docker.internal:9090',
        forcePathStyle: true,
        credentials: { accessKeyId: 'test', secretAccessKey: 'test' },
      }
    : {
        region: 'auto',
        endpoint: process.env.R2_ENDPOINT,
        credentials: {
          accessKeyId: process.env.R2_ACCESS_KEY_ID,
          secretAccessKey: process.env.R2_SECRET_ACCESS_KEY,
        },
      }
);

const BUCKET = process.env.IS_LOCAL
  ? (process.env.S3_BUCKET_NAME || 'my-test-bucket')
  : process.env.R2_BUCKET_NAME;

const streamToBuffer = (stream) =>
  new Promise((resolve, reject) => {
    const chunks = [];
    stream.on('data', (c) => chunks.push(c));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
    stream.on('error', reject);
  });

const getJson = async (key) => {
  const res = await s3Client.send(new GetObjectCommand({ Bucket: BUCKET, Key: key }));
  const buf = await streamToBuffer(res.Body);
  return JSON.parse(buf.toString());
};

const putJson = (key, obj) =>
  s3Client.send(new PutObjectCommand({
    Bucket: BUCKET,
    Key: key,
    Body: JSON.stringify(obj, null, 2),
    ContentType: 'application/json',
  }));

const isNotFound = (err) => err?.name === 'NoSuchKey' || err?.Code === 'NoSuchKey' || err?.$metadata?.httpStatusCode === 404;

// V2 published statics live under the `v2/` key prefix on R2 (bucket root holds V1).
const V2_PREFIX = 'v2/';

export const removeStaticJSON = async (category, grv_id) => {
  const errors = [];

  // 1. index.json — filter out the sensor, recompute stats
  try {
    let index;
    try {
      index = await getJson(`${V2_PREFIX}index.json`);
    } catch (err) {
      if (isNotFound(err)) {
        console.log('index.json not found, skipping');
        index = null;
      } else {
        throw err;
      }
    }
    if (index !== null) {
      index.sensors = (index.sensors ?? []).filter((s) => s.id !== grv_id);
      const allLigands = new Set();
      index.sensors.forEach((s) => (s.ligands ?? []).forEach((l) => allLigands.add(l)));
      index.stats = { regulators: index.sensors.length, ligands: allLigands.size };
      await putJson(`${V2_PREFIX}index.json`, index);
    }
  } catch (err) {
    console.log('Failed to update index.json:', err);
    errors.push(err);
  }

  // 2. indexes/{category}.json — filter out the sensor, recompute count
  try {
    const key = `${V2_PREFIX}indexes/${category.toLowerCase()}.json`;
    let familyIndex;
    try {
      familyIndex = await getJson(key);
    } catch (err) {
      if (isNotFound(err)) {
        console.log(`${key} not found, skipping`);
        familyIndex = null;
      } else {
        throw err;
      }
    }
    if (familyIndex !== null) {
      familyIndex.data = (familyIndex.data ?? []).filter((s) => s.id !== grv_id);
      familyIndex.count = familyIndex.data.length;
      await putJson(key, familyIndex);
    }
  } catch (err) {
    console.log(`Failed to update indexes/${category.toLowerCase()}.json:`, err);
    errors.push(err);
  }

  // 3. sensors/{category}/{grv_id}.json — delete the object
  try {
    const key = `${V2_PREFIX}sensors/${category.toLowerCase()}/${grv_id}.json`;
    try {
      await s3Client.send(new DeleteObjectCommand({ Bucket: BUCKET, Key: key }));
    } catch (err) {
      if (isNotFound(err)) {
        console.log(`${key} not found, nothing to delete`);
      } else {
        throw err;
      }
    }
  } catch (err) {
    console.log(`Failed to delete sensors/${category.toLowerCase()}/${grv_id}.json:`, err);
    errors.push(err);
  }

  // 4. all-sensors.json — filter out the sensor, recompute count and version
  try {
    let all;
    try {
      all = await getJson(`${V2_PREFIX}all-sensors.json`);
    } catch (err) {
      if (isNotFound(err)) {
        console.log('all-sensors.json not found, skipping');
        all = null;
      } else {
        throw err;
      }
    }
    if (all !== null) {
      all.sensors = (all.sensors ?? []).filter((s) => s.id !== grv_id);
      all.count = all.sensors.length;
      all.version = new Date().toISOString();
      await putJson(`${V2_PREFIX}all-sensors.json`, all);
    }
  } catch (err) {
    console.log('Failed to update all-sensors.json:', err);
    errors.push(err);
  }

  if (errors.length > 0) {
    throw new Error(`R2 removal completed with ${errors.length} error(s); see logs for details`);
  }
};

File: functions/deleteSensorV2/s3_remover_v2.py
import datetime
import json
import os

import boto3
import botocore.exceptions

# V2 published statics live under the `v2/` key prefix on R2 (bucket root holds V1).
V2_PREFIX = "v2/"


def _s3_client():
    if os.environ.get("IS_LOCAL"):
        return boto3.client(
            "s3",
            region_name="us-east-2",
            endpoint_url="http://host.docker.internal:9090",
            aws_access_key_id="test",
            aws_secret_access_key="test",
        )
    return boto3.client(
        "s3",
        region_name="auto",
        endpoint_url=os.environ.get("R2_ENDPOINT"),
        aws_access_key_id=os.environ.get("R2_ACCESS_KEY_ID"),
        aws_secret_access_key=os.environ.get("R2_SECRET_ACCESS_KEY"),
    )


def _bucket():
    if os.environ.get("IS_LOCAL"):
        return os.environ.get("S3_BUCKET_NAME") or "my-test-bucket"
    return os.environ.get("R2_BUCKET_NAME")


def _get_json(key):
    client = _s3_client()
    resp = client.get_object(Bucket=_bucket(), Key=key)
    return json.loads(resp["Body"].read())


def _put_json(key, obj):
    client = _s3_client()
    client.put_object(
        Bucket=_bucket(),
        Key=key,
        Body=json.dumps(obj, indent=2),
        ContentType="application/json",
    )


def _delete_object(key):
    client = _s3_client()
    client.delete_object(Bucket=_bucket(), Key=key)


def _is_not_found(err):
    if not isinstance(err, botocore.exceptions.ClientError):
        return False
    code = err.response.get("Error", {}).get("Code")
    if code in ("NoSuchKey", "404"):
        return True
    return err.response.get("ResponseMetadata", {}).get("HTTPStatusCode") == 404


def remove_static_json(category, grv_id):
    errors = []

    # 1. index.json — filter out the sensor, recompute stats
    try:
        try:
            index = _get_json(f"{V2_PREFIX}index.json")
        except Exception as err:
            if _is_not_found(err):
                print("index.json not found, skipping")
                index = None
            else:
                raise
        if index is not None:
            sensors = [s for s in (index.get("sensors") or []) if s.get("id") != grv_id]
            all_ligands = set()
            for s in sensors:
                for ligand in s.get("ligands") or []:
                    all_ligands.add(ligand)
            index["sensors"] = sensors
            index["stats"] = {"regulators": len(sensors), "ligands": len(all_ligands)}
            _put_json(f"{V2_PREFIX}index.json", index)
    except Exception as err:
        print(f"Failed to update index.json: {err}")
        errors.append(err)

    # 2. indexes/{category}.json — filter out the sensor, recompute count
    try:
        key = f"{V2_PREFIX}indexes/{category.lower()}.json"
        try:
            family_index = _get_json(key)
        except Exception as err:
            if _is_not_found(err):
                print(f"{key} not found, skipping")
                family_index = None
            else:
                raise
        if family_index is not None:
            items = [s for s in (family_index.get("data") or []) if s.get("id") != grv_id]
            family_index["data"] = items
            family_index["count"] = len(items)
            _put_json(key, family_index)
    except Exception as err:
        print(f"Failed to update indexes/{category.lower()}.json: {err}")
        errors.append(err)

    # 3. sensors/{category}/{grv_id}.json — delete the object
    try:
        key = f"{V2_PREFIX}sensors/{category.lower()}/{grv_id}.json"
        try:
            _delete_object(key)
        except Exception as err:
            if _is_not_found(err):
                print(f"{key} not found, nothing to delete")
            else:
                raise
    except Exception as err:
        print(f"Failed to delete sensors/{category.lower()}/{grv_id}.json: {err}")
        errors.append(err)

    # 4. all-sensors.json — filter out the sensor, recompute count and version
    try:
        try:
            all_sensors = _get_json(f"{V2_PREFIX}all-sensors.json")
        except Exception as err:
            if _is_not_found(err):
                print("all-sensors.json not found, skipping")
                all_sensors = None
            else:
                raise
        if all_sensors is not None:
            sensors = [s for s in (all_sensors.get("sensors") or []) if s.get("id") != grv_id]
            all_sensors["sensors"] = sensors
            all_sensors["count"] = len(sensors)
            all_sensors["version"] = datetime.datetime.now(datetime.timezone.utc).isoformat()
            _put_json(f"{V2_PREFIX}all-sensors.json", all_sensors)
    except Exception as err:
        print(f"Failed to update all-sensors.json: {err}")
        errors.append(err)

    if errors:
        raise Exception(
            f"R2 removal completed with {len(errors)} error(s); see logs for details"
        )
